### PR TITLE
#159204693 Implement email verification on successful signup

### PR DESCRIPTION
**Pull request by megame24**
#### What does this PR do?
Send verification email to a user on successful signup
#### Description of Task to be completed?
- Send verification email to a user after signup
- Resend verification email on request
#### How should this be manually tested?
- Install the app (`npm install`)
- Start the application with (`npm start`)
- Test POST `/api/users` route, providing username, email and password on postman to recieve verification email on successful sign up.
- Test POST `/api/users/resend-email` route, providing email on postman to resend verification email.
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
#159204693
#### Screenshots (if appropriate)
screen shot 2018-08-02 at 12 31 20 am
#### Questions:
N/A

### CODE REVIEW

File: .env-sample [hunk 1]
```diff
@@ -1,4 +1,15 @@
 PORT=3000
 DATABASE_URL_DEV=postgresql://username:password@host:localhost:5432/database_name
 DATABASE_URL_TEST=postgresql://username:password@host:localhost:5432/database_name
-DATABASE_URL=postgresql://username:password@host:localhost:5432/database_name
```
Comment (mppdg) [inline .env-sample:4]: Why did you remove the third database connection url `DATABASE_URL=postgresql://username:password@host:localhost:5432/database_name` format?

Comment (megame24) [reply .env-sample:4]: Thought it was a repetition, I'll add it now.

File: helpers/verificationHelper.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  /**
+   * Validate email on re-sending verification email
+   * @param {String} email - email passed, on requesting for verification link resend
+   * @returns {Boolean} false
+   */
+  validateEmail: function(email) {
+    if(!email) return { statusCode: 400, data: { status: 'error', message: 'Email not provided' } };
+    if(!/^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()\.,;\s@\"]+\.{0,1})+[^<>()\.,;:\s@\"]{2,})$/.test(email)) {
+      return { statusCode: 400, data: { status: 'error', message: 'Please provide a valid email' } };
```
Comment (mppdg) [inline helpers/verificationHelper.js]: Why is the error message in a `data` property? I think it makes it less intuitive because the `data` property is usually associated with a successful operation.

**Review by mppdg — COMMENTED** [review]
Great job! Your naming scheme like `doNotVerify`, `doNotResendEmail`  makes it easy to understand what the code is doing. The abstractions and organization of modules/functions also enhance readability, maintainability and follow the single responsibility principle.

Don't forget to work on the inline reviews. For consistency, you may want to use the error format on the README.md page i.e
```
{
  "errors":{
    "body": [
      "can't be empty"
    ]
  }
}
```
or as will be established by the team/TTL.

Also, don't forget to squash commits and rebase.

File: .env-sample [hunk 3]
```diff
@@ -1,4 +1,16 @@
...
+GMAIL_PASSWORD=<password>
+
+AUTH_SECRET=<string>
+
+BASEURL=<url>
+
+MAIL_TRAP_HOST=<host>
+MAIL_TRAP_PORT=<port>
+MAIL_TRAP_USER=<userID>
+MAIL_TRAP_PASS=<userPassword>
```
Comment (chidioguejiofor) [inline .env-sample]: Add Newline at end of file

**Review by chidioguejiofor — APPROVED** [review]
Looks good to me. I like your creativity with the tests. 
However, you should correct those new lines

File: test/users.test.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        })
+        .end((err, res) => {
+          expect(res.status).to.equal(200);
+          expect(res.body).to.haveOwnProperty('message')
+            .to.equal('Your verification link has been resent');
+          done();
+        });
+    });
+  });
+});
```
Comment (JCanaks) [inline test/users.test.js:156]: Everything seems fine. Great Job! I will only add that you properly describe how your task can be manually tested in your PR description

Comment (megame24) [reply test/users.test.js:156]: Done

**Review by mppdg — APPROVED** [review]

File: .env-sample (hunk 3 above)
Comment (rhenshaw56) [reply .env-sample]: @megame24 if you fixed this, then reply with done so the reviewer can know you did.

File: package.json [hunk 5]
```diff
@@ -5,9 +5,12 @@
   "main": "index.js",
   "scripts": {
     "start": "node index.js",
-    "test": "cross-env NODE_ENV=test mocha ./test/**/*.js",
-    "coverage": "cross-env NODE_ENV=test ./node_modules/istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha",
-    "lint": "eslint ."
+    "start:dev": "nodemon index.js",
```
Comment (rhenshaw56) [inline package.json]: I see you integrated nodemon already, was it part of the scope of this ticket ?

File: routes/api/users.js [hunk 6]
```diff
@@ -1,87 +1,140 @@
...
-            email: user.email,
-            username: user.username
-        });
-    }).catch(next);
+  User.create({ username, email, hash })
+    .then(function (user) {
+      // return the user and a promise call to send the verification email
+      return { sendMail: verificationHelper.sendVerificationEmail(user), user: user }
+    })
+    .then(function (result) {
```
Comment (rhenshaw56) [inline routes/api/users.js]: use `user` instead of result

File: routes/api/users.js [hunk 7]
```diff
@@ -1,87 +1,140 @@
...
+    .catch(next);
+});
+
+router.get("/users/verify/:token", function (req, res, next) {
+  const token = req.params.token;
+  const decoded = tokenService.verifyToken(token, process.env.AUTH_SECRET);
+  if (!decoded) return res.status(400).json({ errors: { message: "The link has expired" } });
+  User.findById(decoded.id)
+    .then(function (user) {
+      const doNotVerify = verificationHelper.validateUser(user);
```
Comment (rhenshaw56) [inline routes/api/users.js]: change `doNotVerify` to `verificationStatus` or `verifiedStatus` cos when you say `doNotVerify` it means that `validateUser` function will not verify the user

File: routes/api/users.js [hunk 8]
```diff
@@ -1,87 +1,140 @@
...
+        return res.status(doNotVerify.statusCode).json(doNotVerify.error);
+      }
+      user.verified = true;
+      user.save();
+      return res.status(200).json({ status: "success", message: "Your account has been verified" });
+    })
+    .catch(next);
+});
 
+router.post("/users/verify/resend-email", function (req, res, next) {
```
Comment (rhenshaw56) [inline routes/api/users.js]: fix the hound violations

File: routes/api/users.js [hunk 9]
```diff
@@ -1,87 +1,140 @@
...
+  let email = req.body.email;
```
[the remark this answers is not in the document]
Comment (rhenshaw56) [reply routes/api/users.js]: ignore this

File: routes/api/users.js [hunk 10]
```diff
@@ -1,87 +1,140 @@
...
+  email = email ? email.trim() : undefined;
+  const notAValidEmail = verificationHelper.validateEmail(email);
+  if (notAValidEmail) {
+    return res.status(notAValidEmail.statusCode).json(notAValidEmail.error);
+  }
+  User.findOne({ where: { email: email } })
+    .then(function (user) {
+      const doNotResendEmail = verificationHelper.validateUser(user);
```
Comment (rhenshaw56) [inline routes/api/users.js]: don't use `doNot` as a prefix, think of a better name for it

File: seeders/20180801074021-user.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      email: 'naruto@gmail.com',
+      hash: '12345',
+      verified: false,
+      createdAt: new Date(),
+      updatedAt: new Date()
+    }], {});
+  },
+
+  down: (queryInterface, Sequelize) => {
+    /*
```
Comment (rhenshaw56) [inline seeders/20180801074021-user.js]: Take this out

File: services/tokenService.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+module.exports = {
+  /**
+   * Generate a token
+   * @param {Object} payload - object to be encoded into a token
+   * @param {Number} expiresIn - time for token to expire
+   * @returns {String} token
+   */
+  generateToken(payload, expiresIn) {
+    return jwt.sign(payload, process.env.AUTH_SECRET, {expiresIn: expiresIn});
```
[the remark this answers is not in the document]
Comment (rhenshaw56) [reply services/tokenService.js]: fix violations mehn

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

File: .env-sample (hunk 3 above)
Comment (megame24) [reply .env-sample]: @chidioguejiofor done.

File: package.json (hunk 5 above)
Comment (megame24) [reply package.json]: I will remove it.

File: helpers/verificationHelper.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
+/* eslint-disable no-useless-escape */
```
Comment (rhenshaw56) [inline helpers/verificationHelper.js]: get this part out and fix it

File: helpers/verificationHelper.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+   */
+  validateEmail: (email) => {
+    if (!email) {
+      return {
+        status: false,
+        statusCode: 400,
+        error: { errors: { message: 'Email not provided' } },
+      };
+    }
+    /* eslint-disable max-len */
```
Comment (rhenshaw56) [inline helpers/verificationHelper.js]: pls take this bit off

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]
